### PR TITLE
Provides test tool functions to make it easier to write unit tests.

### PR DESCRIPTION
**Pull request by chengxiangdong**
**What this PR does / why we need it**:

Provides test tool functions to make it easier to write unit tests.
  Please see `data_source_huaweicloud_waf_certificate_test.go` and `resource_huaweicloud_waf_certificate_test.go` for usage.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

## PR Checklist

* [x] Tests added/passed.
* [ ] Documentation updated.
* [ ] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/waf' TESTARGS='-run TestAccWafDomainV1'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/waf -v -run TestAccWafDomainV1 -timeout 360m -parallel 4
=== RUN   TestAccWafDomainV1_basic
=== PAUSE TestAccWafDomainV1_basic
=== RUN   TestAccWafDomainV1_policy
=== PAUSE TestAccWafDomainV1_policy
=== CONT  TestAccWafDomainV1_basic
=== CONT  TestAccWafDomainV1_policy
--- PASS: TestAccWafDomainV1_policy (64.69s)
--- PASS: TestAccWafDomainV1_basic (104.95s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/waf       105.045s

make testacc TEST='./huaweicloud/services/acceptance/waf' TESTARGS='-run TestAccWafCertificateV1_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/waf -v -run TestAccWafCertificateV1_basic -timeout 360m -parallel 4
=== RUN   TestAccWafCertificateV1_basic
=== PAUSE TestAccWafCertificateV1_basic
=== CONT  TestAccWafCertificateV1_basic
--- PASS: TestAccWafCertificateV1_basic (85.28s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/waf       85.596s
```

### CODE REVIEW

File: huaweicloud/services/acceptance/acceptance.go [hunk 1]
```diff
@@ -76,3 +82,179 @@ func TestAccPreCheckEpsID(t *testing.T) {
 func RandomAccResourceName() string {
 	return fmt.Sprintf("tf_acc_test_%s", acctest.RandString(5))
 }
+
+const (
+	// UNSET check if the resource is not set, corresponding to: TestCheckNoResourceAttrSet.
+	UNSET = "#NOSET"
```
Comment (Lance52259) [inline huaweicloud/services/acceptance/acceptance.go]: Use Camel-Case for constant naming

Comment (chengxiangdong) [reply huaweicloud/services/acceptance/acceptance.go]: Obsolete.

File: huaweicloud/services/acceptance/acceptance.go [hunk 2]
```diff
@@ -45,6 +51,184 @@ func init() {
...
+func InitResourceCheck(resourceName string, resourceObject interface{}, getResourceFunc ServiceFunc) *resourceCheck {
+	return &resourceCheck{
+		resourceName:    resourceName,
+		resourceObject:  resourceObject,
+		getResourceFunc: getResourceFunc,
+		resourceType:    resourceTypeCode,
+	}
+}
+
+const checkAttrRegexpStr = `^\$\{([^\}]+)\}$`
```
Comment (ShiChangkuo) [inline huaweicloud/services/acceptance/acceptance.go]: please put all const variables together

File: huaweicloud/services/acceptance/acceptance.go [hunk 3]
```diff
@@ -45,6 +51,184 @@ func init() {
 	}
 }
 
+// ServiceFunc the HuaweiCloud resource query functions.
+type ServiceFunc func(*config.Config, *terraform.ResourceState) (interface{}, error)
+
+// resourceCheck resource check object, only used in the package.
+type resourceCheck struct {
+	resourceName    string
+	checkMap        map[string]string
```
Comment (ShiChangkuo) [inline huaweicloud/services/acceptance/acceptance.go]: as we discussed, `checkMap` is not used any more, so we can delete it.

File: huaweicloud/services/acceptance/acceptance.go [hunk 4]
```diff
@@ -45,6 +51,184 @@ func init() {
...
+		return resName, keyName, fmtp.Errorf("The type of 'variable' is error, "+
+			"expected ${resourceType.name.field} got %s", varStr)
+	}
+
+	reg, err := regexp.Compile(checkAttrRegexpStr)
+	if err != nil {
+		return resName, keyName, fmtp.Errorf("The acceptance function is wrong.")
+	}
+	mArr := reg.FindStringSubmatch(varStr)
+	if len(mArr) < 1 {
```
Comment (ShiChangkuo) [inline huaweicloud/services/acceptance/acceptance.go]: the length of mArr is always >= 1

File: huaweicloud/services/acceptance/acceptance.go [hunk 5]
```diff
@@ -45,6 +51,184 @@ func init() {
...
+			}
+		}
+
+		for _, rs := range s.RootModule().Resources {
+			if rs.Type != resourceType {
+				continue
+			}
+
+			conf := TestAccProvider.Meta().(*config.Config)
+			if _, err := rc.getResourceFunc(conf, rs); err == nil {
```
Comment (ShiChangkuo) [inline huaweicloud/services/acceptance/acceptance.go]: should check `getResourceFunc` is nil or not?

File: huaweicloud/services/acceptance/waf/resource_huaweicloud_waf_certificate_test.go [hunk 6]
```diff
@@ -17,25 +17,37 @@ func TestAccWafCertificateV1_basic(t *testing.T) {
...
 
+	rc := acceptance.InitResourceCheck(
+		resourceName,
+		&certificate,
+		func(c *config.Config, state *terraform.ResourceState) (interface{}, error) {
+			wafClient, err := c.WafV1Client(acceptance.HW_REGION_NAME)
+			if err != nil {
+				return nil, fmt.Errorf("error creating HuaweiCloud WAF client: %s", err)
+			}
+			return certificates.Get(wafClient, state.Primary.ID).Extract()
```
Comment (ShiChangkuo) [inline huaweicloud/services/acceptance/waf/resource_huaweicloud_waf_certificate_test.go]: suggest define it as a new function.

File: huaweicloud/services/acceptance/acceptance.go (hunk 2 above)
Comment (chengxiangdong) [reply huaweicloud/services/acceptance/acceptance.go]: fixed

File: huaweicloud/services/acceptance/acceptance.go (hunk 4 above)
Comment (chengxiangdong) [reply huaweicloud/services/acceptance/acceptance.go]: fixed

File: huaweicloud/services/acceptance/acceptance.go (hunk 5 above)
Comment (chengxiangdong) [reply huaweicloud/services/acceptance/acceptance.go]: added

File: huaweicloud/services/acceptance/waf/resource_huaweicloud_waf_certificate_test.go (hunk 6 above)
Comment (chengxiangdong) [reply huaweicloud/services/acceptance/waf/resource_huaweicloud_waf_certificate_test.go]: fixed

File: huaweicloud/services/acceptance/acceptance.go (hunk 3 above)
Comment (chengxiangdong) [reply huaweicloud/services/acceptance/acceptance.go]: deleted

**Pull request merged**